### PR TITLE
Bump TypeScript and related dependencies

### PR DESCRIPTION
**Pull request by Mrtenz**
This bumps the following dependencies:

- `typescript` from `~5.0.4` to `~5.3.3`.
- `@typescript-eslint/parser` from `^5.42.1` to `^6.21.0`.
- `eslint-plugin-jsdoc` from `^39.6.2` to `^41.1.2`.
- `@types/readable-stream` from `^2.3.15` to `^4.0.15`.

The bumps to the related dependencies are necessary because some APIs in TypeScript/ESLint were deprecated and removed.

### CODE REVIEW

**Comment by Mrtenz** [comment]
@metamaskbot update-pr

File: package.json [hunk 1]
```diff
@@ -104,8 +104,7 @@
     "semver": "^7.5.4",
     "simple-git-hooks": "^2.7.0",
     "ts-node": "^10.9.1",
-    "tsup": "^8.0.1",
```
Comment (FrederikBolding) [inline package.json:107]: Premature deletion? 😄

Comment (Mrtenz) [reply package.json:107]: Messed up a rebase. Fixed now!

File: packages/snaps-controllers/src/types/vendor/global.d.ts [hunk 2]
```diff
@@ -1,8 +0,0 @@
-export {};
-
-// eslint-disable-next-line import/unambiguous
-declare global {
```
Comment (FrederikBolding) [inline packages/snaps-controllers/src/types/vendor/global.d.ts:4]: How come the vendor types aren't needed anymore?

File: packages/snaps-controllers/src/interface/SnapInterfaceController.ts [hunk 3]
```diff
@@ -199,7 +200,7 @@ export class SnapInterfaceController extends BaseController<
       // possibly infinite.
       draftState.interfaces[id] = {
         snapId,
-        content: element,
+        content: castDraft(element),
```
Comment (FrederikBolding) [inline packages/snaps-controllers/src/interface/SnapInterfaceController.ts:203]: Explain please!

File: packages/examples/packages/wasm/.eslintrc.js [hunk 4]
```diff
@@ -1,28 +1,17 @@
-const { resolve } = require('path');
-
 module.exports = {
   extends: ['../../.eslintrc.js'],
 
   parserOptions: {
     tsconfigRootDir: __dirname,
   },
 
-  overrides: [
```
Comment (FrederikBolding) [inline packages/examples/packages/wasm/.eslintrc.js:10]: ?

File: packages/snaps-controllers/src/types/vendor/global.d.ts (hunk 2 above)
Comment (Mrtenz) [reply packages/snaps-controllers/src/types/vendor/global.d.ts:4]: `DecompressionStream` is now included in TypeScript's on lib files, and the latest version of `@types/readable-stream` has types for `Duplex` and `pipeline`.

File: packages/snaps-controllers/src/interface/SnapInterfaceController.ts (hunk 3 above)
Comment (Mrtenz) [reply packages/snaps-controllers/src/interface/SnapInterfaceController.ts:203]: "`SnapElement` is not assignable to `WritableDraft<SnapElement>`" or something like that.

File: packages/examples/packages/wasm/.eslintrc.js (hunk 4 above)
Comment (FrederikBolding) [reply packages/examples/packages/wasm/.eslintrc.js:10]: So should we remove the tsconfig entirely from the WASM code?

Comment (Mrtenz) [reply packages/examples/packages/wasm/.eslintrc.js:10]: I'm not sure but I can if you prefer?

**Review by FrederikBolding — APPROVED** [review]

**Pull request merged**